### ISSUE TITLE
segfault padding with memor config

### ISSUE THREAD

**Issue by mert-kurttutan** [mert-kurttutan/issue-buffer#38]
```python
#!/usr/bin/env python3

from __future__ import annotations

import argparse

import torch

import ttnn


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Minimal repro for sharded memory config tile-alignment failure.")
    parser.add_argument("--batch", type=int, default=1)
    parser.add_argument("--length", type=int, default=3560)
    parser.add_argument("--channels", type=int, default=192)
    parser.add_argument("--device-id", type=int, default=0)
    return parser.parse_args()


def main() -> None:
    args = parse_args()

    device = ttnn.open_device(device_id=args.device_id)

    torch_x = torch.randn((args.batch, args.length, args.channels), dtype=torch.bfloat16)
    x = ttnn.from_torch(
        torch_x,
        dtype=ttnn.bfloat16,
        layout=ttnn.TILE_LAYOUT,
        device=device,
    )

    grid_size = device.compute_with_storage_grid_size()
    core_grid = ttnn.CoreGrid(y=grid_size.y, x=grid_size.x)
    candidate_grids = [
        ttnn.CoreGrid(y=grid_size.y, x=1),
        ttnn.CoreGrid(y=1, x=grid_size.x),
        ttnn.CoreGrid(y=1, x=1),
    ]
    strategy = ttnn.ShardStrategy.HEIGHT

    input_length = x.shape[1]
    pad_length = 32 - (input_length % 32) if (input_length % 32) != 0 else 0
    x_padded_length = input_length + pad_length
    x_padded_shape = (x.shape[0], x_padded_length, x.shape[2])

    print(f"device grid size: {grid_size}")
    print(f"primary core_grid: {core_grid}")
    print(f"candidate_grids: {candidate_grids}")
    print(f"x shape before padding: {x.shape}")
    print(f"x padded shape: {x_padded_shape}")
    print(f"strategy: {strategy}")

    memory_config = ttnn.create_sharded_memory_config_(
        ttnn.Shape([x.shape[0], x_padded_length, x.shape[2]]),
        core_grid=core_grid,
        strategy=strategy,
        orientation=ttnn.ShardOrientation.ROW_MAJOR,
    )

    print(f"memory_config: {memory_config}")

    if pad_length > 0:
        x = ttnn.pad(x, [(0, 0), (0, pad_length), (0, 0)], value=0.0, memory_config=memory_config)
        print(f"x shape after pad: {x.shape}")

    ttnn.close_device(device)


if __name__ == "__main__":
    main()

```
